### PR TITLE
update for input stream refcount

### PR DESCRIPTION
**Pull request by TingDaoK**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samples/s3/s3-cp.c [hunk 1]
```diff
@@ -551,12 +551,21 @@ static int s_kickoff_put_object(
...
+        .vtable = &s_update_input_stream_vtable,
+        .impl = update_stream,
+        .allocator = cp_app_ctx->app_ctx->allocator,
+    };
+
+    struct aws_input_stream *wrap_stream = aws_input_stream_new(&input_stream_options);
+    aws_http_message_set_body_stream(request_options.message, wrap_stream);
 
     transfer_ctx->meta_request = aws_s3_client_make_meta_request(cp_app_ctx->app_ctx->client, &request_options);
+    /* message will owns the stream */
```
Comment (cmello) [inline samples/s3/s3-cp.c:566]: nitpick suggestion: `/* message owns the stream */`

File: samples/s3/s3-cp.c [hunk 2]
```diff
@@ -233,42 +233,42 @@ int s_input_seek(struct aws_input_stream *stream, int64_t offset, enum aws_strea
...
+    aws_mem_release(update_stream->allocator, update_stream);
 }
 
 static struct aws_input_stream_vtable s_update_input_stream_vtable = {
     .get_length = s_input_get_length,
     .seek = s_input_seek,
     .read = s_input_read,
     .get_status = s_input_get_status,
-    .destroy = s_input_destroy,
+    .impl_destroy = s_input_destroy,
```
Comment (cmello) [inline samples/s3/s3-cp.c:271]: Are we renaming `.destroy` to `.impl_destroy`, or keeping both? Not sure if the `impl_` prefix makes it more readable from the interface perspective.

**Review by cmello — APPROVED** [review]
LGTM

File: samples/s3/s3-cp.c (hunk 2 above)
Comment (TingDaoK) [reply samples/s3/s3-cp.c:271]: we are renaming it. So, for now the input stream will be clean up when the refcount drops to zero, and when it happens the `impl_destroy` will be invoked to clean up the impl passed in.

File: source/s3_chunk_stream.c [hunk 3]
```diff
@@ -138,38 +144,40 @@ static int s_aws_input_chunk_stream_read(struct aws_input_stream *stream, struct
...
         status->is_valid = true;
         return AWS_OP_SUCCESS;
     }
-    int err = aws_input_stream_get_status(impl->current_stream, status);
-    if (!err) {
+    int res = aws_input_stream_get_status(impl->current_stream, status);
+    if (res != AWS_OP_SUCCESS) {
+        /* Only when the current_stream is NULL, it is end of stream, as the current stream will be updated to feed to
+         * data */
         status->is_end_of_stream = false;
```
Comment (graebm) [inline source/s3_chunk_stream.c:158]: trivial: this comment doesn't help me understand what's going on

**Review by graebm — APPROVED** [review]
fix & ship

**Pull request merged**